doc/developer/design/20210910_include_kafka_metadata.md

```diff
@@ -0,0 +1,130 @@
+# Include Kafka Metadata
+
+## Summary
+
+Users have regularly requested access to certain pieces of Kafka (and other)
+per-message metadata within views. The most common requests have been for the
+timestamp and offset fields.
+
+The proposal here is to extend our syntax to allow users to opt-in to including
+message headers on a per-source basis.
+
+## Goals
+
+Give users easy access to commonly-requested metadata without any backwards
+compatibility concerns.
+
+## Non-goals
+
+Generic support for Headers. The Kafka API allows users to specify [arbitrary
+per-message headers][headers-javadoc]. As far as I know we haven't received
+requests for this yet, and the Kafka API does not provide enough information for
+us to automatically decode header values (it's [just bytes][headers-rustdoc]),
+so the exact shape of what we would give to users requires more design work.
+
+[headers-javadoc]: https://kafka.apache.org/27/javadoc/org/apache/kafka/connect/header/Headers.html
+[headers-rustdoc]: https://docs.rs/rdkafka/0.26.0/rdkafka/message/trait.Headers.html
+
+## Description
+
+We will extend the existing `INCLUDE` syntax to take an optional `METADATA
+(timestamp | offset | partition | topic)+` clause. These are all the remaining
+fields in a Kafka message that are not `headers`. Each field inside the field
+list can take an optional `AS <name>` clause, in case of conflicts with the key
+or value names.
+
+With this, the full syntax for the `INCLUDE` clause becomes:
+
+```
+INCLUDE (
+    KEY (AS <name>)?
+    | METADATA ((timestamp | offset | partition | topic) (AS <name>)?)+
+)
+```
+
+With an example of just the new syntax looking like:
+
+`INCLUDE METADATA (timestamp, offset)`
+
+And an example of it being combined with existing uses of `INCLUDE KEY` full
+syntax:
+
+`INCLUDE KEY, METADATA (timestamp, offset, partition, topic)`
+
+and with the renaming syntax:
+
+`INCLUDE KEY AS mykey, METADATA (timestamp AS ts, offset as "Off")`
```

Comment:
I agree with Frank that this is missing a paragraph about what happens. (I mean, it's pretty obvious, I think, but would be good to say.) E.g.

> "If `TIMESTAMP` is specified, it appears as an additional column in the source of type `int` that contains..."